v2/pkg/operators/matchers/validate.go

```diff
@@ -0,0 +1,54 @@
+package matchers
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/projectdiscovery/sliceutil"
+	"gopkg.in/yaml.v3"
+)
+
+var commonExpectedFields = []string{"type", "condition", "name", "match-all", "negative"}
+
+// Validate perform initial validation on the matcher structure
+func (matcher *Matcher) Validate() error {
+	// uses yaml marshaling to convert the struct to map[string]interface to have same field names
+	matcherMap := make(map[string]interface{})
+	marshaledMatcher, err := yaml.Marshal(matcher)
+	if err != nil {
+		return err
+	}
+	if err := yaml.Unmarshal(marshaledMatcher, &matcherMap); err != nil {
+		return err
+	}
+
+	var expectedFields []string
+	switch matcher.matcherType {
+	case DSLMatcher:
+		expectedFields = append(commonExpectedFields, "dsl")
+	case StatusMatcher:
+		expectedFields = append(commonExpectedFields, "status", "part")
+	case SizeMatcher:
+		expectedFields = append(commonExpectedFields, "size", "part")
+	case WordsMatcher:
+		expectedFields = append(commonExpectedFields, "words", "part", "encoding", "case-insensitive")
+	case BinaryMatcher:
+		expectedFields = append(commonExpectedFields, "binary", "part", "encoding", "case-insensitive")
```

Comment:
It would be MUCH nicer if we could do this without hard-coded strings.

Reply:
Maybe differentiating into separate matchers with a common embeddable struct containing common fields could be an option. But it would require a lot of refactoring and additional complexity. This seems a reasonable compromise even if the fields list needs a manual update.

Reply:
Would it be possible to reference the field names directly and then use reflection to request their tag names?

This way at least it would be refactor safe.